In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.156
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4400.31
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Overwriting APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

4 8
12

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Overwriting sum.py


In [ ]:
!python sum.py

2 5
7


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

5
1 2 3 4 5
20


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Writing input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Writing input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



----------------------------------------------------**Test *3.1***-------------------------------------------------

In [ ]:
%%writefile input4.txt
10
7 8 9554  5214  42  567 331 44  4 1

Writing input4.txt


In [ ]:
9554*5214

49814556

In [ ]:
%%writefile ctest4.txt
49814556

Writing ctest4.txt


In [ ]:
!./max_pairwise_product_long_full < input4.txt

49814556


In [ ]:
%%writefile output4.txt
49814556

Writing output4.txt


In [ ]:
!./max_pairwise_product_long_full < input4.txt > ctest4.txt
!grep -xvFf output4.txt ctest4.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

----------------------------------------------------Test 3.2-------------------------------------------------

In [ ]:
%%writefile input5.txt
5
23  5 4455  6871  25

Writing input5.txt


In [ ]:
4455*6871

30610305

In [ ]:
%%writefile ctest5.txt
30610305

Writing ctest5.txt


In [ ]:
!./max_pairwise_product_long_full < input5.txt

30610305


In [ ]:
%%writefile output5.txt
30610305

Writing output5.txt


In [ ]:
!./max_pairwise_product_long_full < input5.txt > ctest5.txt
!grep -xvFf output5.txt ctest5.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

---------------------------------------------------Test 3.3-------------------------------------------------

In [ ]:
%%writefile input6.txt
20
12  3 6 7 918 4 1 2 85  3 7 23  45  68  72  23  100 58  75  64

Writing input6.txt


In [ ]:
918*100

91800

In [ ]:
%%writefile ctest6.txt
91800

Writing ctest6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt

91800


In [ ]:
%%writefile output6.txt
91800

Writing output6.txt


In [ ]:
!./max_pairwise_product_long_full < input6.txt > ctest6.txt
!grep -xvFf output6.txt ctest6.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

----------------------------------------------------Test 3.4-------------------------------------------------

In [ ]:
%%writefile input7.txt
3
233356464  112745  4556658

Writing input7.txt


In [ ]:
233356464*4556658

1063325598537312

In [ ]:
%%writefile ctest7.txt
1063325598537312

Writing ctest7.txt


In [ ]:
!./max_pairwise_product_long_full < input7.txt

1063325598537312


In [ ]:
%%writefile output7.txt
1063325598537312

Writing output7.txt


In [ ]:
!./max_pairwise_product_long_full < input7.txt > ctest7.txt
!grep -xvFf output7.txt ctest7.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

---------------------------------------------------Test 3.5-------------------------------------------------

In [ ]:
%%writefile input8.txt
4
10101  10001101  101 1

Writing input8.txt


In [ ]:
10101*10001101

101021121201

In [ ]:
%%writefile ctest8.txt
101021121201

Writing ctest8.txt


In [ ]:
!./max_pairwise_product_long_full < input8.txt

101021121201


In [ ]:
%%writefile output8.txt
101021121201

Writing output8.txt


In [ ]:
!./max_pairwise_product_long_full < input8.txt > ctest8.txt
!grep -xvFf output8.txt ctest8.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

---------------------------------------------------Test 3.6-------------------------------------------------

In [ ]:
%%writefile input9.txt
7
2 3 5 7 4 9 1

Writing input9.txt


In [ ]:
7*9

63

In [ ]:
%%writefile ctest9.txt
63

Writing ctest9.txt


In [ ]:
!./max_pairwise_product_long_full < input9.txt

63


In [ ]:
%%writefile output9.txt
63

Writing output9.txt


In [ ]:
!./max_pairwise_product_long_full < input9.txt > ctest9.txt
!grep -xvFf output9.txt ctest9.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

---------------------------------------------------Test 3.7-------------------------------------------------

In [ ]:
%%writefile input10.txt
12
2 3 5 7 4 9 1 55 44 56 78 98

Writing input10.txt


In [ ]:
78*98

7644

In [ ]:
%%writefile ctest10.txt
7644

Writing ctest10.txt


In [ ]:
!./max_pairwise_product_long_full < input10.txt

7644


In [ ]:
%%writefile output10.txt
7644

Writing output10.txt


In [ ]:
!./max_pairwise_product_long_full < input10.txt > ctest10.txt
!grep -xvFf output10.txt ctest10.txt

In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

40000000000

real	5m6.795s
user	4m59.238s
sys	0m0.217s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	5m4.108s
user	4m57.506s
sys	0m0.172s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Writing max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

109



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input4.txt
n= 10


For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


200001000021


In [ ]:
# Estimated time by instruction

306/200001000010


1.5299923499617505e-09

In [ ]:
# Estimated number of instructions by second

306/200001000010

1.5299923499617505e-09

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.873s
user	0m0.852s
sys	0m0.003s


## Python

In [ ]:
# Crear y escribir en el archivo
with open("input13.txt", "w") as file:
    file.write("11000\n")  # n
    for i in range(1, 11001):
        file.write(f"{i} ")

print("Archivo input13.txt creado y escrito correctamente.")


Archivo input13.txt creado y escrito correctamente.


In [ ]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0

    for first in range(len(numbers)):
        for second in range(len(numbers)):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    M = max_pairwise_product(numbers)
    print(M)


Overwriting max_pairwise_product_python.py


In [ ]:
# Verify
!time python max_pairwise_product_python.py < input13.txt

120989000

real	0m21.137s
user	0m20.803s
sys	0m0.023s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


Writing max_pairwise_product_long.cpp


In [ ]:
%%writefile max_pairwise_product_python.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])

    return max_product

if __name__ == "__main__":
    n = int(input())
    numbers = list(map(int, input().split()))

    print(max_pairwise_product(numbers))


In [ ]:
%%writefile max_pairwise_product_long.py
def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(first + 1, n):
            aux_product = numbers[first] * numbers[second]
            if max_product < aux_product:
                max_product = aux_product

    return max_product


### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

**-------------------------------------------------------------------------------C++-------------------------------------------------------------------------**

In [ ]:
%%writefile introsort.cpp

#include <iostream>
#include <vector>
#include <algorithm>

// Function to count the number of instructions executed
int instruction_count = 0;

// Custom comparator function for introsort
bool custom_compare(int a, int b) {
    // Count a comparison instruction
    instruction_count++;
    return a < b;
}

int main() {
    int n;
    std::cin >> n; // Read the number of elements

    std::vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        std::cin >> arr[i]; // Read the elements
    }

    std::sort(arr.begin(), arr.end(), custom_compare);

    // Print the sorted array
    for (int num : arr) {
        std::cout << num << " ";
    }

    std::cout << std::endl;

    // Print the number of instructions executed
    std::cout << "Instruction Count: " << instruction_count << std::endl;

    return 0;
}


Writing introsort.cpp


**-----------------------------------------------------------------Python---------------------------------------------------------------------**

In [ ]:
%%writefile timsort.py

def timsort(arr):
    # Count the number of instructions executed
    instruction_count = 0

    arr.sort()

    return instruction_count

if __name__ == "__main__":
    n = int(input())
    arr = list(map(int, input().split()))

    # Sort the array using timsort and get the instruction count
    instructions = timsort(arr)

    # Print the sorted array
    print("Sorted Array:", arr)

    # Print the number of instructions executed
    print("Instruction Count:", instructions)


Writing timsort.py


Test with inputs 1 and 2



In [ ]:
!g++ introsort.cpp -o introsort
!./introsort < input1.txt

1 2 3 4 5 6 7 8 9 10 
Instruction Count: 18


In [ ]:
!g++ introsort.cpp -o introsort
!./introsort < input2.txt

1 2 3 4 5 6 7 8 9 10 
Instruction Count: 9


In [ ]:
!python timsort.py < input1.txt

Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Instruction Count: 0


In [ ]:
!python timsort.py < input2.txt

Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Instruction Count: 0


In [ ]:
!python timsort.py < input2.txt

Sorted Array: [1, 2, 5, 5, 6, 9]
Instruction Count: 0


Test and time it with input 3

In [ ]:
!python timsort.py < input3.txt

Sorted Array: [1, 3, 9, 23, 200000, 200000]
Instruction Count: 0


In [ ]:
!g++ introsort.cpp -o introsort
!./introsort < input3.txt

1 3 9 23 200000 200000 
Instruction Count: 13


20 stress tests

creamos 20 inputs del 14 al 34

In [ ]:
import random

for i in range(14, 35):
    filename = f"input{i}.txt"
    n = random.randint(1, 23)  # Número de elementos en cada archivo

    with open(filename, "w") as file:
        file.write(f"{n}\n")  # Escribir el valor de n
        numbers = list(range(1, n + 1))
        random.shuffle(numbers)  # Mezclar los números en orden aleatorio
        file.write(" ".join(map(str, numbers)))  # Escribir los números en el archivo

    print(f"Archivo {filename} creado y escrito correctamente.")


Archivo input14.txt creado y escrito correctamente.
Archivo input15.txt creado y escrito correctamente.
Archivo input16.txt creado y escrito correctamente.
Archivo input17.txt creado y escrito correctamente.
Archivo input18.txt creado y escrito correctamente.
Archivo input19.txt creado y escrito correctamente.
Archivo input20.txt creado y escrito correctamente.
Archivo input21.txt creado y escrito correctamente.
Archivo input22.txt creado y escrito correctamente.
Archivo input23.txt creado y escrito correctamente.
Archivo input24.txt creado y escrito correctamente.
Archivo input25.txt creado y escrito correctamente.
Archivo input26.txt creado y escrito correctamente.
Archivo input27.txt creado y escrito correctamente.
Archivo input28.txt creado y escrito correctamente.
Archivo input29.txt creado y escrito correctamente.
Archivo input30.txt creado y escrito correctamente.
Archivo input31.txt creado y escrito correctamente.
Archivo input32.txt creado y escrito correctamente.
Archivo inpu

Count executed instructions in code and verify with inputs 1,2 and 3


c++

In [ ]:

 %%writefile torture_introsort.sh
#!/bin/bash
g++ introsort.cpp -o introsort
for i in {14..34}
do
    echo "Running introsort with input$i.txt"
    ./introsort < input$i.txt
done

Overwriting torture_introsort.sh


In [ ]:
!chmod u+x torture_introsort.sh
!./torture_introsort.sh

Running introsort with input14.txt
1 2 3 4 5 6 7 8 9 10 11 
Instruction Count: 36
Running introsort with input15.txt
1 2 3 
Instruction Count: 2
Running introsort with input16.txt
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 
Instruction Count: 82
Running introsort with input17.txt
1 2 3 4 5 6 7 8 9 
Instruction Count: 21
Running introsort with input18.txt
1 2 3 4 5 6 7 8 9 
Instruction Count: 24
Running introsort with input19.txt
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 
Instruction Count: 116
Running introsort with input20.txt
1 2 3 4 5 6 7 8 9 10 11 12 13 
Instruction Count: 40
Running introsort with input21.txt
1 2 3 4 5 6 7 8 9 10 
Instruction Count: 27
Running introsort with input22.txt
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 
Instruction Count: 86
Running introsort with input23.txt
1 2 3 4 5 6 7 
Instruction Count: 17
Running introsort with input24.txt
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 
Instruction Count: 79
Running introsort with input25.txt
1 2 

PYTHON

In [ ]:

 %%writefile torture_timsort.sh
#!/bin/bash

for i in {14..34}
do
  echo "Running timsort with input$i.txt"
  python timsort.py < input$i.txt
done

In [ ]:
!chmod u+x torture_timsort.sh
!./torture_timsort.sh

Running timsort with input14.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
Instruction Count: 0
Running timsort with input15.txt
Sorted Array: [1, 2, 3]
Instruction Count: 0
Running timsort with input16.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
Instruction Count: 0
Running timsort with input17.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9]
Instruction Count: 0
Running timsort with input18.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9]
Instruction Count: 0
Running timsort with input19.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]
Instruction Count: 0
Running timsort with input20.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
Instruction Count: 0
Running timsort with input21.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Instruction Count: 0
Running timsort with input22.txt
Sorted Array: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
Instruction Cou

Calculate estimated time by instruction and number of instructions by second

c++

In [ ]:

 %%writefile torture_time_introsort.sh
#!/bin/bash
g++ introsort.cpp -o introsort
for i in {14..34}
do
    echo "Running introsort with input$i.txt"
    (time ./introsort < input$i.txt) 2>&1 | grep -E 'real|sys'
done


Writing torture_time_introsort.sh


In [ ]:
!chmod u+x  torture_time_introsort.sh
!./torture_time_introsort.sh

Running introsort with input14.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input15.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input16.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input17.txt
real	0m0.003s
sys	0m0.001s
Running introsort with input18.txt
real	0m0.003s
sys	0m0.001s
Running introsort with input19.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input20.txt
real	0m0.003s
sys	0m0.003s
Running introsort with input21.txt
real	0m0.003s
sys	0m0.001s
Running introsort with input22.txt
real	0m0.004s
sys	0m0.002s
Running introsort with input23.txt
real	0m0.004s
sys	0m0.000s
Running introsort with input24.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input25.txt
real	0m0.003s
sys	0m0.000s
Running introsort with input26.txt
real	0m0.003s
sys	0m0.002s
Running introsort with input27.txt
real	0m0.003s
sys	0m0.003s
Running introsort with input28.txt
real	0m0.004s
sys	0m0.001s
Running introsort with input29.txt
real	0m0.003s
sys	0m0.002s
Running 

python

In [ ]:
 %%writefile torture_time_timsort.sh
#!/bin/bash
for i in {14..34}
do
    echo "Running timsort with input$i.txt"
    (time python timsort.py < input$i.txt) 2>&1 | grep -E 'real|sys'
done


In [ ]:
!chmod u+x torture_time_timsort.sh
!./torture_time_timsort.sh

Running timsort with input14.txt
real	0m0.047s
sys	0m0.011s
Running timsort with input15.txt
real	0m0.046s
sys	0m0.008s
Running timsort with input16.txt
real	0m0.046s
sys	0m0.012s
Running timsort with input17.txt
real	0m0.044s
sys	0m0.014s
Running timsort with input18.txt
real	0m0.048s
sys	0m0.012s
Running timsort with input19.txt
real	0m0.057s
sys	0m0.012s
Running timsort with input20.txt
real	0m0.056s
sys	0m0.015s
Running timsort with input21.txt
real	0m0.044s
sys	0m0.009s
Running timsort with input22.txt
real	0m0.045s
sys	0m0.013s
Running timsort with input23.txt
real	0m0.053s
sys	0m0.007s
Running timsort with input24.txt
real	0m0.060s
sys	0m0.010s
Running timsort with input25.txt
real	0m0.062s
sys	0m0.016s
Running timsort with input26.txt
real	0m0.055s
sys	0m0.014s
Running timsort with input27.txt
real	0m0.046s
sys	0m0.012s
Running timsort with input28.txt
real	0m0.051s
sys	0m0.008s
Running timsort with input29.txt
real	0m0.050s
sys	0m0.008s
Running timsort with input30.txt
real	0m

### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

c++

In [ ]:
%%writefile find_cpp.cpp
#include <iostream>
#include <vector>
#include <climits> // Include the <climits> header for INT_MIN

void findLargestAndSecondLargest(const std::vector<int>& arr) {
    int largest = INT_MIN;          // Initialize with smallest possible integer
    int secondLargest = INT_MIN;    // Initialize with smallest possible integer

    for (int num : arr) {
        if (num > largest) {
            secondLargest = largest;
            largest = num;
        } else if (num > secondLargest && num != largest) {
            secondLargest = num;
        }
    }

    std::cout << "Largest: " << largest << std::endl;
    std::cout << "Second Largest: " << secondLargest << std::endl;
}

int main() {
    int n;
    std::cin >> n;  // Read the number of elements

    std::vector<int> arr(n);

    for (int i = 0; i < n; i++) {
        std::cin >> arr[i];  // Read the elements
    }

    findLargestAndSecondLargest(arr);

    return 0;
}


Writing find_cpp.cpp


PYTHON

In [ ]:
%%writefile find_python.py
def find_largest_and_second_largest(arr):
    largest = float('-inf')  # Initialize with negative infinity
    second_largest = float('-inf')  # Initialize with negative infinity

    for num in arr:
        if num > largest:
            second_largest = largest
            largest = num
        elif num > second_largest and num != largest:
            second_largest = num

    print("Largest:", largest)
    print("Second Largest:", second_largest)

if __name__ == "__main__":
    n = int(input())  # Read the number of elements
    arr = list(map(int, input().split()))  # Read the elements

    find_largest_and_second_largest(arr)


Overwriting find_python.py


Test with inputs 1 and 2



C++

In [ ]:
!g++ find_cpp.cpp -o find_cpp
!./find_cpp < input1.txt

Largest: 10
Second Largest: 9


In [ ]:
!./find_cpp < input2.txt

Largest: 10
Second Largest: 9


PYTHON

In [ ]:
!python find_python.py < input1.txt

Largest: 10
Second Largest: 9


In [ ]:
!python find_python.py < input2.txt

Largest: 10
Second Largest: 9


Test and time it with input 3

In [ ]:
!g++ find_cpp.cpp -o find_cpp
!./find_cpp < input3.txt

Largest: 200000
Second Largest: 23


In [ ]:
!python find_python.py < input3.txt

Largest: 200000
Second Largest: 23


20 stress tests

In [ ]:

 %%writefile torture_find_cpp.sh
#!/bin/bash
g++  find_cpp.cpp -o find_cpp
for i in {14..34}
do
    echo "Running find_cpp with input$i.txt"
    ./find_cpp < input$i.txt
done

Writing torture_find_cpp.sh


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
!chmod u+x torture_find_cpp.sh
!./torture_find_cpp.sh

Running find_cpp with input14.txt
Largest: 11
Second Largest: 10
Running find_cpp with input15.txt
Largest: 3
Second Largest: 2
Running find_cpp with input16.txt
Largest: 19
Second Largest: 18
Running find_cpp with input17.txt
Largest: 9
Second Largest: 8
Running find_cpp with input18.txt
Largest: 9
Second Largest: 8
Running find_cpp with input19.txt
Largest: 22
Second Largest: 21
Running find_cpp with input20.txt
Largest: 13
Second Largest: 12
Running find_cpp with input21.txt
Largest: 10
Second Largest: 9
Running find_cpp with input22.txt
Largest: 19
Second Largest: 18
Running find_cpp with input23.txt
Largest: 7
Second Largest: 6
Running find_cpp with input24.txt
Largest: 16
Second Largest: 15
Running find_cpp with input25.txt
Largest: 7
Second Largest: 6
Running find_cpp with input26.txt
Largest: 8
Second Largest: 7
Running find_cpp with input27.txt
Largest: 2
Second Largest: 1
Running find_cpp with input28.txt
Largest: 6
Second Largest: 5
Running find_cpp with input29.txt
Largest:

python


In [ ]:

 %%writefile torture_find_python.sh
#!/bin/bash

for i in {14..34}
do
  echo "Running find_python with input$i.txt"
  python find_python.py < input$i.txt
done

Writing torture_find_python.sh


In [ ]:
!chmod u+x torture_find_python.sh
!./torture_find_python.sh

Running find_python with input14.txt
Largest: 11
Second Largest: 10
Running find_python with input15.txt
Largest: 3
Second Largest: 2
Running find_python with input16.txt
Largest: 19
Second Largest: 18
Running find_python with input17.txt
Largest: 9
Second Largest: 8
Running find_python with input18.txt
Largest: 9
Second Largest: 8
Running find_python with input19.txt
Largest: 22
Second Largest: 21
Running find_python with input20.txt
Largest: 13
Second Largest: 12
Running find_python with input21.txt
Largest: 10
Second Largest: 9
Running find_python with input22.txt
Largest: 19
Second Largest: 18
Running find_python with input23.txt
Largest: 7
Second Largest: 6
Running find_python with input24.txt
Largest: 16
Second Largest: 15
Running find_python with input25.txt
Largest: 7
Second Largest: 6
Running find_python with input26.txt
Largest: 8
Second Largest: 7
Running find_python with input27.txt
Largest: 2
Second Largest: 1
Running find_python with input28.txt
Largest: 6
Second Largest:

Calculate estimated time by instruction and number of instructions by second

C++

In [ ]:
 %%writefile torture_time_find_cpp.sh
#!/bin/bash
g++ introsort.cpp -o introsort
for i in {14..34}
do
    echo "Running time´s find_cpp with input$i.txt"
    (time ./introsort < input$i.txt) 2>&1 | grep -E 'real|sys'
done

Overwriting torture_time_find_cpp.sh


In [ ]:
!chmod u+x  torture_time_find_cpp.sh
!./torture_time_find_cpp.sh

Running time´s find_cpp with input14.txt
real	0m0.003s
sys	0m0.002s
Running time´s find_cpp with input15.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input16.txt
real	0m0.002s
sys	0m0.001s
Running time´s find_cpp with input17.txt
real	0m0.002s
sys	0m0.000s
Running time´s find_cpp with input18.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input19.txt
real	0m0.002s
sys	0m0.001s
Running time´s find_cpp with input20.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input21.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input22.txt
real	0m0.002s
sys	0m0.001s
Running time´s find_cpp with input23.txt
real	0m0.002s
sys	0m0.001s
Running time´s find_cpp with input24.txt
real	0m0.002s
sys	0m0.000s
Running time´s find_cpp with input25.txt
real	0m0.002s
sys	0m0.000s
Running time´s find_cpp with input26.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input27.txt
real	0m0.002s
sys	0m0.002s
Running time´s find_cpp with input28.txt
real	0m

PYTHON


In [ ]:
 %%writefile torture_time_find_python.sh
#!/bin/bash
for i in {14..34}
do
    echo "Running time´s find_python with input$i.txt"
    (time python find_python < input$i.txt) 2>&1 | grep -E 'real|sys'
done

Overwriting torture_time_find_python.sh


In [ ]:
!chmod u+x  torture_time_find_python.sh
!./torture_time_find_python.sh

Running time´s find_python with input14.txt
real	0m0.043s
sys	0m0.008s
Running time´s find_python with input15.txt
real	0m0.057s
sys	0m0.011s
Running time´s find_python with input16.txt
real	0m0.053s
sys	0m0.010s
Running time´s find_python with input17.txt
real	0m0.047s
sys	0m0.010s
Running time´s find_python with input18.txt
real	0m0.048s
sys	0m0.007s
Running time´s find_python with input19.txt
real	0m0.059s
sys	0m0.013s
Running time´s find_python with input20.txt
real	0m0.048s
sys	0m0.010s
Running time´s find_python with input21.txt
real	0m0.053s
sys	0m0.010s
Running time´s find_python with input22.txt
real	0m0.047s
sys	0m0.009s
Running time´s find_python with input23.txt
real	0m0.047s
sys	0m0.009s
Running time´s find_python with input24.txt
real	0m0.044s
sys	0m0.008s
Running time´s find_python with input25.txt
real	0m0.051s
sys	0m0.012s
Running time´s find_python with input26.txt
real	0m0.053s
sys	0m0.014s
Running time´s find_python with input27.txt
real	0m0.049s
sys	0m0.012s
Runnin

### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile find_cpp_op.cpp
#include <iostream>
#include <climits>

int main() {
    int largest = INT_MIN;          // Initialize largest with smallest possible integer
    int secondLargest = INT_MIN;    // Initialize secondLargest with smallest possible integer
    int num;                        // Variable to store the current number

    while (std::cin >> num) {       // Read numbers from input until EOF
        // Check if the current number is greater than the largest
        // If yes, update both largest and secondLargest
        if (num > largest) {
            secondLargest = largest;
            largest = num;
        }
        // Check if the current number is greater than the secondLargest
        // If yes, update only the secondLargest
        else if (num > secondLargest && num != largest) {
            secondLargest = num;
        }
    }

    std::cout << "Largest: " << largest << std::endl;
    std::cout << "Second Largest: " << secondLargest << std::endl;

    return 0;
}


Writing find_cpp_op.cpp


Test with inputs 1 and 2




In [ ]:
!g++ find_cpp_op.cpp -o find_cpp_op
!./find_cpp_op < input1.txt

Largest: 10
Second Largest: 9


In [ ]:
!./find_cpp_op < input2.txt

Largest: 10
Second Largest: 9


Test and time it with input 3

In [ ]:
!./find_cpp_op < input3.txt

Largest: 200000
Second Largest: 23


20 stress tests

In [ ]:

 %%writefile torture_find_cpp_op.sh
#!/bin/bash
g++  find_cpp_op.cpp -o find_cpp_op
for i in {14..34}
do
    echo "Running find_cpp with input$i.txt"
    ./find_cpp_op < input$i.txt
done

Overwriting torture_find_cpp_op.sh


In [ ]:
!chmod u+x  torture_find_cpp_op.sh
!./torture_find_cpp_op.sh

Running find_cpp with input14.txt
Largest: 11
Second Largest: 10
Running find_cpp with input15.txt
Largest: 3
Second Largest: 2
Running find_cpp with input16.txt
Largest: 19
Second Largest: 18
Running find_cpp with input17.txt
Largest: 9
Second Largest: 8
Running find_cpp with input18.txt
Largest: 9
Second Largest: 8
Running find_cpp with input19.txt
Largest: 22
Second Largest: 21
Running find_cpp with input20.txt
Largest: 13
Second Largest: 12
Running find_cpp with input21.txt
Largest: 10
Second Largest: 9
Running find_cpp with input22.txt
Largest: 19
Second Largest: 18
Running find_cpp with input23.txt
Largest: 7
Second Largest: 6
Running find_cpp with input24.txt
Largest: 16
Second Largest: 15
Running find_cpp with input25.txt
Largest: 7
Second Largest: 6
Running find_cpp with input26.txt
Largest: 8
Second Largest: 7
Running find_cpp with input27.txt
Largest: 2
Second Largest: 1
Running find_cpp with input28.txt
Largest: 6
Second Largest: 5
Running find_cpp with input29.txt
Largest:

Count executed instructions in code and verify with inputs 1,2 and 3

c++

In [ ]:
!echo "instructions for input1.txt"
!objdump -d ./find_cpp_op | wc -l < input1.txt


instructions for input1.txt
2


In [ ]:
!echo "instructions for input3.txt"
!objdump -d ./find_cpp_op | wc -l < input2.txt


instructions for input3.txt
2


In [ ]:
%%writefile instructions_cpp.cpp
#include <iostream>
#include <climits>

int main() {
    int largest = INT_MIN;          // 1: Initialize largest with smallest possible integer
    int secondLargest = INT_MIN;    // 1: Initialize secondLargest with smallest possible integer
    int num;                        // 1: Variable to store the current number
    int instruction_count = 3;      // 3 instructions for variable initialization

    while (std::cin >> num) {       // Read numbers from input until EOF
        // Check if the current number is greater than the largest
        // If yes, update both largest and secondLargest
        // 3: Comparison, 2 assignments, 1 arithmetic operation (num > largest)
        if (num > largest) {
            secondLargest = largest;    // 1: Assignment
            largest = num;              // 1: Assignment
        }
        // Check if the current number is greater than the secondLargest
        // If yes, update only the secondLargest
        // 3: Comparison, 1 assignment (num > secondLargest)
        // 1: Comparison (num != largest)
        else if (num > secondLargest && num != largest) {
            secondLargest = num;        // 1: Assignment
        }
        instruction_count += 8;  // Update instruction count for each loop iteration
    }
    std::cout << "Instruction Count: " << instruction_count << std::endl;  // Print total instruction count

    return 0;
}


Overwriting instructions_cpp.cpp


In [ ]:
!g++ instructions_cpp.cpp -o instructions_cpp
!./instructions_cpp < input1.txt

Instruction Count: 91


In [ ]:
!./instructions_cpp < input2.txt

Instruction Count: 91


In [ ]:
!./instructions_cpp < input3.txt

Instruction Count: 59


Calculate estimated time by instruction and number of instructions by second

In [ ]:
%%writefile estimated.cpp
#include <iostream>
#include <climits>
#include <chrono>

int main() {
    int largest = INT_MIN;
    int secondLargest = INT_MIN;
    int num;
    int instruction_count = 3;

    auto start_time = std::chrono::high_resolution_clock::now();

    while (std::cin >> num) {
        if (num > largest) {
            secondLargest = largest;
            largest = num;
        }
        else if (num > secondLargest && num != largest) {
            secondLargest = num;
        }
        instruction_count += 8;
    }

    auto end_time = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed_seconds = end_time - start_time;
    double execution_time = elapsed_seconds.count();

    std::cout << "Instruction Count: " << instruction_count << std::endl;
    std::cout << "Execution Time (seconds): " << execution_time << std::endl;

    // Calculate estimated time per instruction
    double estimated_time_per_instruction = execution_time / instruction_count;

    // Calculate instructions per second
    double instructions_per_second = 1.0 / estimated_time_per_instruction;

    std::cout << "Estimated Time per Instruction (seconds): " << estimated_time_per_instruction << std::endl;
    std::cout << "Instructions per Second: " << instructions_per_second << std::endl;

    return 0;
}


Writing estimated.cpp


In [ ]:
!g++ estimated.cpp -o estimated
!./estimated < input1.txt

Instruction Count: 91
Execution Time (seconds): 8.485e-05
Estimated Time per Instruction (seconds): 9.32418e-07
Instructions per Second: 1.07248e+06


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

Overwriting find_python_op.py


Test with inputs 1 and 2




Test and time it with input 3

20 stress tests

Count executed instructions in code and verify with inputs 1,2 and 3

Calculate estimated time by instruction and number of instructions by second